Implement keystone authentication webhook

alkurbatov opened the pull request:
This pull request brings implementation of keystone authentication webhook.

Please pay attention that build process is clumsy now (one has to enter a lot of data manually). That will be changed in the nearest future.

Commit:
Implement keystone authentication webhook

ibazhitov commented on Makefile:
```diff
@@ -0,0 +1,11 @@
+TARGET := build/$(BINARY)
+
+$(TARGET):
+	glide install
```
Better add '--strip-vendor' flag to remove nested vendor directories after install.

alkurbatov replied:
Thank you, fixed in a separate commit.

Commit:
bld: Remove nested vendor directories

Use 'glide install' with the '--strip-vendor' option
to remove nested vendor directories.

avagin commented on cmd/main.go:
```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Are you sure about "The Kubernetes Authors"?

alkurbatov replied:
We plan try to merge this webhook to upstream in the nearest future. Until it is done I'll replace the copyrights in a separate commit which could be easily reverted in future.

avagin commented on cmd/main.go:
```diff
@@ -0,0 +1,60 @@
+	defer glog.Flush()
+
+	if *certFile == "" || *certKey == "" {
+		flag.Usage()
```
It is better to write a reason why it failed

alkurbatov replied:
Will do.

avagin commented on cmd/main.go:
```diff
@@ -0,0 +1,60 @@
+	http.Handle("/authenticate", newAuthWebhook())
+
+	glog.Infof("Serving on :%d", *port)
+	glog.Fatal(server.ListenAndServeTLS(*certFile, *certKey))
```
glog.Fatal() always exits with 255, doesn't it?  Does it mean that this script always fail?

I don't like when something useful is executed from log functions. It looks messy. I would suggest to rewrite this code by this way:
```
if err := server.ListenAndServeTLS(*certFile, *certKey): err != nil {
    glog.Fatal(err)
}
```

alkurbatov replied:
I've just copy-pasted it from official docs of net/http :) In general http server serves forever unless Ctrl+C is pressed or some fatal error occurred.

avagin commented on cmd/auth.go:
```diff
@@ -0,0 +1,78 @@
+	resp := &Payload{}
+	err := json.NewDecoder(req.Body).Decode(resp)
+	if err != nil {
+		glog.Error(err)
```
I think it is better to write more informative messages.
flog.Errorf("Unable to decode a request body: %v", err)

alkurbatov replied:
Will do.

avagin commented:
Did you think about tests? I can recommend to use travic-ci for basic checks: compilation, golint, etc.

alkurbatov commented:
Yep we need the tests and travis especially for another repo (feathers-authentication-keystone). However there is no travis yet. I need to figure out how it works etc.

I believe the tests will be introduced in a separate pull-request.

Commit:
Change copyrights

Replace k8s copyrights with our own. This commit should be reverted
when we will send this code to the upstream.

Commit:
Show better error when no cert or key is provided

Commit:
Improve serverside error reporting

Form meaningful messages on various errors.
